io/io/inc/ROOT/RIoUring.hxx

```diff
@@ -0,0 +1,63 @@
+/*************************************************************************
+ * Copyright (C) 1995-2020, Rene Brun and Fons Rademakers.               *
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * For the licensing terms see $ROOTSYS/LICENSE.                         *
+ * For the list of contributors see $ROOTSYS/README/CREDITS.             *
+ *************************************************************************/
+
+#ifndef ROOT_RIoUring
+#define ROOT_RIoUring
+
+#include <cstring>
+#include <liburing.h>
+#include <liburing/io_uring.h>
+
+#include <ROOT/RError.hxx>
+#include <ROOT/RLogger.hxx>
+
+using ROOT::Experimental::RException;
+
+namespace ROOT {
+namespace Internal {
+
+class RIoUring {
+private:
+   struct io_uring fRing;
+public:
+   explicit RIoUring(size_t size) {
+      int ret = io_uring_queue_init(size, &fRing, 0 /* no flags */);
+      if (ret) {
+         throw RException(R__FAIL("Error initializing io_uring: " + std::string(std::strerror(-ret))));
+      }
+   }
+
+   RIoUring(const RIoUring&) = delete;
+   RIoUring& operator=(const RIoUring&) = delete;
+
+   ~RIoUring() {
+      io_uring_queue_exit(&fRing);
+   }
+
+   /// Check if io_uring is available on this system.
+   static bool IsAvailable() {
+      thread_local bool couldOpenRing = false;
```

Comment:
We could do something like

    static bool result = CheckIfAvailable();
    return result;

CheckIfAvailable will be called at most once and in a thread-safe manner.

Reply:
Thank you, this way is much nicer :) (41858029cba8cc7c55d4bf9ae1440f8269fe091a)